scripts/py_matter_yamltests/matter_yamltests/constraints.py

```diff
@@ -199,169 +362,440 @@ def check_response(self, value, value_type_name) -> bool:
+
+        return f'The response ({value}) is missing {expected_values}.'
+
 
 class _ConstraintExcludes(BaseConstraint):
-    def __init__(self, excludes):
-        super().__init__(types=[list])
+    def __init__(self, context, excludes):
+        super().__init__(context, types=[list])
         self._excludes = excludes
 
     def check_response(self, value, value_type_name) -> bool:
         return set(self._excludes).isdisjoint(value)
 
+    def get_reason(self, value, value_type_name) -> str:
+        unexpected_values = []
+
+        for unexpected_value in self._excludes:
+            if unexpected_value in value:
+                unexpected_values.append(unexpected_value)
+
+        return f'The response ({value}) contains {unexpected_values}.'
+
 
 class _ConstraintHasMaskSet(BaseConstraint):
-    def __init__(self, has_masks_set):
-        super().__init__(types=[int])
+    def __init__(self, context, has_masks_set):
+        super().__init__(context, types=[int])
         self._has_masks_set = has_masks_set
 
     def check_response(self, value, value_type_name) -> bool:
         return all([(value & mask) == mask for mask in self._has_masks_set])
 
+    def get_reason(self, value, value_type_name) -> str:
+        expected_masks = []
+
+        for expected_mask in self._has_masks_set:
+            if (value & expected_mask) != expected_mask:
+                expected_masks.append(hex(expected_mask))
+
+        return f'The response ({hex(value)}) does not match the masks: {expected_masks}.'
+
 
 class _ConstraintHasMaskClear(BaseConstraint):
-    def __init__(self, has_masks_clear):
-        super().__init__(types=[int])
+    def __init__(self, context, has_masks_clear):
+        super().__init__(context, types=[int])
         self._has_masks_clear = has_masks_clear
 
     def check_response(self, value, value_type_name) -> bool:
         return all([(value & mask) == 0 for mask in self._has_masks_clear])
 
+    def get_reason(self, value, value_type_name) -> str:
+        unexpected_masks = []
+
+        for unexpected_mask in self._has_masks_clear:
+            if (value & unexpected_mask) == unexpected_mask:
+                unexpected_masks.append(hex(unexpected_mask))
+
+        return f'The response ({hex(value)}) match the masks: {unexpected_masks}.'
+
 
 class _ConstraintNotValue(BaseConstraint):
-    def __init__(self, not_value):
-        super().__init__(types=[], is_null_allowed=True)
+    def __init__(self, context, not_value):
+        super().__init__(context, types=[], is_null_allowed=True)
         self._not_value = not_value
 
     def check_response(self, value, value_type_name) -> bool:
         return value != self._not_value
 
+    def get_reason(self, value, value_type_name) -> str:
+        return f'The response value "{value}" should differs from the constraint.'
+
 
 def get_constraints(constraints: dict) -> list[BaseConstraint]:
     _constraints = []
+    context = constraints
```

Comment:
This is correct. We are sharing the same dictionary and if multiple exceptions are raised it won't looks good. The issue with coping the dictionary is that it will then be disconnected from the top level "context" and it will loose some informations.
In general I agree that we should do better. For example the fact that we are altering the "responses" and operates on modified values kill the original informations. Ideally we would carry over the original line from the yaml test file and show this to the user but that would requires some amount of work that does not fit with this PR.
I suggest we keep `untag` for now since it is not hurting but we need to find a plan to give more insights about the error source.

Reply:
SGTM